File: src/main/java/com/wesker926/leetcode/lcof/p030/s1/Solution.java
package com.wesker926.leetcode.lcof.p030.s1;

import java.util.LinkedList;

/**
 * @author wesker.gh
 * @date 2021/10/26
 * @description stack
 */
public class Solution {

    /**
     * Your MinStack object will be instantiated and called as such:
     * MinStack obj = new MinStack();
     * obj.push(x);
     * obj.pop();
     * int param_3 = obj.top();
     * int param_4 = obj.min();
     */
    static class MinStack {

        private final LinkedList<Integer> val;

        private final LinkedList<Integer> min;

        /**
         * initialize your data structure here.
         */
        public MinStack() {
            val = new LinkedList<>();
            min = new LinkedList<>();
        }

        public void push(int x) {
            val.push(x);
            min.push(min.isEmpty() || x <= min.peekFirst() ? x : min.peekFirst());
        }

        public void pop() {
            val.pop();
            min.pop();
        }

        public int top() {
            return val.isEmpty() ? -1 : val.peekFirst();
        }

        public int min() {
            return min.isEmpty() ? -1 : min.peekFirst();
        }
    }
}

File: src/main/java/com/wesker926/leetcode/lcof/p030/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/10/26
// @description: stack
package s1

type MinStack struct {
	cnt int
	val []int
	min []int
}

/** initialize your data structure here. */
func Constructor() MinStack {
	return MinStack{
		cnt: 0,
		val: make([]int, 0),
		min: make([]int, 0),
	}
}

func (this *MinStack) Push(x int) {
	this.val = append(this.val, x)
	if this.cnt == 0 || this.min[this.cnt-1] >= x {
		this.min = append(this.min, x)
	} else {
		this.min = append(this.min, this.min[this.cnt-1])
	}
	this.cnt++
}

func (this *MinStack) Pop() {
	if this.cnt > 0 {
		this.cnt--
		this.val = this.val[:this.cnt]
		this.min = this.min[:this.cnt]
	}
}

func (this *MinStack) Top() int {
	if this.cnt > 0 {
		return this.val[this.cnt-1]
	}
	return 0
}

func (this *MinStack) Min() int {
	if this.cnt > 0 {
		return this.min[this.cnt-1]
	}
	return 0
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.Min();
 */
